Updated mappings for the Hercules Inpulse 300

BoredGuy1 opened the pull request:
I made some improvements to Mixxx's included Inpulse 300 mapping. I tried to stay close to the manufacturer mappings listed in [this manual](https://ts.hercules.com/download/sound/manuals/DJC_Inpulse300/DJControl_Inpulse_300_User_Manual_EN.pdf).

Full list of changes:
- Added ability to stop samplers (shift + button)

- Added toneplay
    - Pressing a pad jumps to the most recently used hotcue and changes the pitch
    - Shift + toneplay changes the pitch without jumping
    - Pads 1-4 increase the pitch by +0 to +3 semitones
    - Pads 5-8 decrease the pitch by -4 to -1 semitones
    - LED reflects total adjustment to current key

- Added slicer/slicer loop:
    - Press slicer button to activate a slicer section, press again to deactivate it
    - Slicer length is dependent on loop length
    - Exiting slicer mode or pressing LOOP OUT also deactivates the slicer section
    - Creating a slicer section overrides active loops, and creating a 4-beat loop overrides active slicer sections
    - Only works on constant BPM (also true in native software DJUCED)
    - Slicer section does not move backwards when scratching (also true in native software DJUCED)

- Added actual beatmatch guide functionality to the LEDs
    - This replaces the LEDs' old end-of-track warning functionality. Hopefully no one missed it :P

- Changed the way scratching works
    - Wheels have inertia rather than instantly stopping when released (this allows backspins and other tricks)
    - Removed scratchPad in XML (not sure what it was doing there)

- Updated VU meter syntax
    - Replaced vu_meter with VuMeter to comply with new version of Mixxx
    - Replaced connectControl with makeConnection to allow multiple connections (and also to comply with newer versions)

- Replaced "hotcue_X_enabled" with "hotcue_X_status" in XML file to comply with newer version of Mixxx

Manual is being updated accordingly in another PR: [PR#604](https://github.com/mixxxdj/manual/pull/604).

BoredGuy1 committed:
Added sampler stopping and toneplay

BoredGuy1 committed:
Edited authors

BoredGuy1 committed:
Edited to-do

BoredGuy1 committed:
Fixed syntax

JoergAtGithub commented:
Welcome at Mixxx!
As a first-time contributor we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

JoergAtGithub commented:
The pre-commit check is failing. The best way to fix pre-commit issues is to install pre-commit locally on your system, as described here: https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking
Than it will fix these issues automatically before every commit.
Alternatively, you can download the pre-commit.patch file from the artifacts of this PR
![grafik](https://github.com/mixxxdj/mixxx/assets/64457745/c3076931-5c8e-4389-b464-f6b4622140b3)
unzip it, and apply it using:
`git apply pre-commit.patch`

Please note, that this will not fix the typo, that the codespell check reports - this one must be fixed manually.

BoredGuy1 committed:
Fixed syntax (again)

BoredGuy1 committed:
Fixed syntax (yet again)

BoredGuy1 committed:
More fixed syntax

BoredGuy1 committed:
More fixed syntax

BoredGuy1 committed:
Replaced let with var (for some reason let causes errors)

BoredGuy1 committed:
Finally fixed spelling (thanks pre-commit)

BoredGuy1 commented:
Alright, I've filled out the agreement, installed pre-commit, and fixed all the typos and style issues. Let me know if there's anything else I need to do.

JoergAtGithub commented:
Target for this PR should be 2.4 not main. Please rebase.

JoergAtGithub commented:
If you need help with Git, have a look here: https://github.com/mixxxdj/mixxx/wiki/Using-Git#targeting-another-base-branch

BoredGuy1 committed:
Added JS to toneplay

BoredGuy1 committed:
Changed scratching, fixed LED issue

BoredGuy1 committed:
Changed the way toneplay LED works

BoredGuy1 committed:
Added slicer/slicer loop

BoredGuy1 committed:
Added shift+toneplay controls

BoredGuy1 committed:
Fixed VU meters

BoredGuy1 committed:
Added beatmatch guide

BoredGuy1 committed:
Fixed scratching, rearranged shift+toneplay functions

BoredGuy1 commented:
I'm still figuring my way around git, but assuming I rebased it properly, it should be ready to merge now. I also took the opportunity to add a whole bunch of other stuff, but now the manual PR is no longer up to date. I will be busy in the coming weeks, but hopefully I can get the manual up to speed after that.

In the meantime, it would be nice to have some feedback on this mapping (especially from @DJPhatso ;)

DJPhatso commented:
Pretty busy right now, but I'll see what I can do.

DJPhatso commented:
Some feedback of what I was able to try:

- **Samplers stop**: 

Work as expected.

- **Toneplay**: 

Been a while since I used this functionality on Traktor, so I might not be the best to judge if the flow is correct, but all functions were assigned as described. 

- **Beatmatch guide to LEDs**

Glad you could make it work as intended :-)

- **Changed the way scratching works**

I noticed some unexpected scratching on occasion (i.e. without actually touching the wheels), which did no occur when I switched back to the default mapping (even tried with another unit to rule out some hardware problems), so there might be some adjustments to be done on that side.

Looking at the script, there are a few User Options I might try to modify next time. 

- **Updated VU meter syntax**

No problems there

BoredGuy1 commented:
Thanks for the feedback!

I just used the script at a small house gig, and I can confirm the scratching issue exists (where the engine activates scratching briefly without jogwheel input). I will need to look into that.

In regards to the shift+pad mode, I think reset key/sync key/move keyboard pads are good to have (although they could be rearranged). But I can see how the pitch up/pitch down functions can be hard to use. I've only used Mixxx and DJUCED, so I'm not sure how other softwares do it.

I was looking at a CDJ the other day - it has a button that, when pressed, changes the functionality of the jogwheel so that it tighten a loop instead of scratching. Would it be more intuitive if I did something like that, but for pitch? I could make it so that when you hold down a toneplay+shift pad, the jogwheel enters some sort of "pitch mode" where you can rotate the jogwheel to pitch. I could also get rid of the shift+toneplay mappings entirely, but I feel like I'm in too far to give up now :P

BoredGuy1 committed:
Rearranged shift+toneplay functions

BoredGuy1 committed:
Rearranged shift+toneplay functions

BoredGuy1 committed:
Fixed slicer

ronso0 commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -3,7 +3,7 @@
 // ***************************************************************************
 // * Mixxx mapping script file for the Hercules DJControl Inpulse 300.
 // * Author: DJ Phatso, contributions by Kerrick Staley and BoredGuy1
-// * Version 1.3 (Feb 2024)
```
Makes sense to update the version, too, no?
(this is no review, just stumbled over this)

BoredGuy1 replied:
Interesting. It's up-to-date on my end. This is what I see:
![image](https://github.com/mixxxdj/mixxx/assets/66702733/d3fd3958-b39a-4140-bc6a-c0bd0ad05402)

I will try to iron out the formatting issues though. Turns out I didn't have ESLint installed, so all the formatting checks were automatically being passed. Oops...

BoredGuy1 committed:
Fixed formatting for real thanks to ESLint

BoredGuy1 commented:
For some reason, the Ubuntu Qt 6.2 build always fails. That seems strange - the script doesn't involve Qt at all, does it?
![image](https://github.com/mixxxdj/mixxx/assets/66702733/ad6bd893-b18f-4b4a-8243-a9e2b68335b1)

BoredGuy1 committed:
Fixed formatting and ESLint (hopefully for the last time)

BoredGuy1 committed:
Changed how toneplay + shift works

BoredGuy1 committed:
Changed XML to account for loop hotcues

BoredGuy1 committed:
Fixed XML version (which I broke 6 months ago)

BoredGuy1 commented:
Never mind, I figured out the QT issue.

This should be ready for review. The manual is also done, sans the broken link checks.

BoredGuy1 commented:
Never mind, the random scratching is still there.

BoredGuy1 committed:
Fixed scratching issue FINALLY (hopefully for real)

BoredGuy1 committed:
Fixed scratching issue for real

BoredGuy1 commented:
Scratching issue has been resolved. Ready for review again.

ywwg commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
+DJCi300.padModeBeatjump = 7;
+
+// Slicer connections
+let slicerSizeConnection1;
```
all variables must be part of the DJCi300 object to prevent possible collisions in javascript

Swiftb0y replied:
The only suggestion I would like to make here is to put these into an array rather than postfixing the variables with numbers:
```suggestion
const slizerConnections = {
   size: [undefined, undefined],
   beat: [undefined, undefined],
   track: [undefined, undefined],
}
```

ywwg commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
+    // Connect the VUMeters
+    engine.makeConnection("[Channel1]", "VuMeter", DJCi300.vuMeterUpdateDeck);
+    engine.makeConnection("[Channel2]", "VuMeter", DJCi300.vuMeterUpdateDeck);
+    engine.makeConnection("[Master]", "VuMeterL", DJCi300.vuMeterUpdateMaster);
```
we use Main now instead of Master

ywwg commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
+    // This saves resources
+    engine.makeConnection("[Channel1]", "beat_distance", DJCi300.updateBeatmatchAlignLED);
+    // We also want to update all beatmatch LEDs when a song is played
+    engine.makeConnection("[Channel1]", "play", DJCi300.updateBeatmatchAlignLED);
```
does this need to be two separate connections to two functions? why not one function that does both things?

BoredGuy1 replied:
I wanted to separate the beat align and tempo update functions. That way, if the tempo is synced but the beats aren't aligned, the program doesn't have to keep checking the tempo. In theory, this is computationally cheaper, although I don't know by how much. 

`updateBeatmatchTempoLED` is connected to the tempo of both decks, since if either deck changes BPM, it could cause the decks to desync.
`updateBeatmatchAlignLED` is only connected to the beat distance of one deck, since it just has to execute on every tick.

I suppose another implementation would be to create a giant beatmatch update function, but it would still have to be connected to `bpm` and `play` on both decks, as well as `beat_distance` on one deck, with the performance downside I mentioned earlier (unless I'm missing something).

ywwg commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -185,11 +381,362 @@ DJCi300.scratchWheel = function(channel, control, value, status, _group) {
+DJCi300.changeMode = function(channel, control, value, _status, _group) {
+    const deck = channel;
+    const oldPadMode = DJCi300.padMode[deck];
+    DJCi300.padMode[deck] = control - 15;
```
what's the 15 here?

BoredGuy1 replied:
Just an offset. For some reason, the MIDI codes for changing pad modes starts at 15 (0x0F), so I subtract 15 to make Mode 1 = 0.

But in hindsight, it would be more readable if I just changed the pad mode definitions at the top.

ywwg commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
+    engine.makeConnection("[Channel2]", "pitch", DJCi300.updateToneplayLED);
+
+    // Define slicer connections by connecting them
+    DJCi300.connectSlicerFunctions(1);
+    DJCi300.connectSlicerFunctions(2);
+    // Then disconnect them since we're not in slicer mode (yet)
+    DJCi300.disconnectSlicerFunctions(1);
+    DJCi300.disconnectSlicerFunctions(2);
```
I think a better model here would be to have the functions always connected, and then use a bool to decide if they are active.  So the handler functions would say "if slicing is not active, return".  Then you don't need to be making and disconnecting connections all the time.

Swiftb0y replied:
That results in lots of function call overhead from C++ into the JS. I think we need to carefully consider how many of these are active vs disabled. (dis-)Connecting is not that expensive IIRC. its just a hashmap lookup with a `QObject::connect` call.

ywwg reviewed (changes requested):
Some initial questions, not a full review.  Thank you for doing this!

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
 DJCi300.init = function() {
     // Scratch button state
-    DJCi300.scratchButtonState = true;
+    DJCi300.scratchButtonState = {
+        1: true,
+        2: true
+    };
```
why this over an array or better names?
```suggestion
    DJCi300.scratchButtonState = {
        "[Channel1]": true,
        "[Channel2]": true
    };
```
```suggestion
    DJCi300.scratchButtonState = [true, true];
```

BoredGuy1 replied:
I originally did it this way so that you can use 1 and 2 directly (based on the decks) instead of 0 and 1. But yeah, I think it would be easier to just use `[Channel1]` and `[Channel2]`.

Swiftb0y replied:
both of the patterns I suggested are established, so choose whatever you prefer.

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
+    };
+    // Slicer buttons (stores whether or not slicer button is pressed)
+    DJCi300.slicerButtonEnabled = {
+        1: [0, 0, 0, 0, 0, 0, 0, 0],
+        2: [0, 0, 0, 0, 0, 0, 0, 0]
```
these are booleans right? Then use those please:
```suggestion
        1: [false, false, false, false, false, false, false, false],
        2: [false, false, false, false, false, false, false, false]
```

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
+    engine.softTakeoverIgnoreNextValue("[Channel2]", "rate");
+
+    // Connect the VUMeters
+    engine.makeConnection("[Channel1]", "VuMeter", DJCi300.vuMeterUpdateDeck);
+    engine.makeConnection("[Channel2]", "VuMeter", DJCi300.vuMeterUpdateDeck);
```
consider deduplicating all of these:
```suggestion
    for (const group of ["[Channel1]","[Channel2]"]) {
        engine.makeConnection(group, "VuMeter", DJCi300.vuMeterUpdateDeck);
        // add the others here as well
    } 
```

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -124,40 +303,57 @@ DJCi300._convertWheelRotation = function(value) {
+// This is called immediately after the wheel is released and we want to switch between scratching and jogging gracefully.
+// It is also connected to callbacks and called regularly to see if the wheel has slowed down enough. Once it does, then we switch from scratching to jogging
+DJCi300.updateScratchAction = function(value, group, _control) {
+    const deck = (group === "[Channel1]") ? 1 : 2;
```
there is `script.deckFromGroup` for this.

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -185,11 +381,362 @@ DJCi300.scratchWheel = function(channel, control, value, status, _group) {
+};
+
+// Mode buttons
+DJCi300.changeMode = function(channel, control, value, _status, _group) {
+    const deck = channel;
```
this would be a good opportunity to incorporate `ComponentJS`. You can probably copy paste large parts of the code there from the Roland DJ-505 mapping (which I personally consider a gold standard componentJS mapping).

BoredGuy1 committed:
Replaced vuMeterUpdateMaster with vuMeterUpdateMain

BoredGuy1 committed:
Changed pad mode values and changed slicerButtons to boolean

BoredGuy1 committed:
Consolidated connections with for statement

BoredGuy1 committed:
changed recentHotcue

BoredGuy1 committed:
changed VU meter syntax

BoredGuy1 committed:
changed hotcue_X_enabled to hotcue_X_status

BoredGuy1 committed:
Replaced stuff with script.deckFromGroup

BoredGuy1 committed:
Replaced separate slicer variables with dictionary declaration

BoredGuy1 committed:
Merge branch 'main' of https://github.com/BoredGuy1/mixxx

ywwg commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -65,25 +65,17 @@ DJCi300.kScratchActionScratch = 1;
-
-DJCi300.vuMeterUpdateMaster = function(value, _group, _control) {
+DJCi300.padModeNone = 0;
+DJCi300.padModeHotcue = 15;
```
Add a comment saying that these numbers correspond directly to the midi control values

BoredGuy1 committed:
Replaced some more numbers with booleans

ywwg commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -44,76 +60,233 @@ DJCi300.scratchShiftMultiplier = 4;
+    midi.sendShortMsg(0x97, 0x48, 0x7F);
+
+    // Connect master VUMeters
+    engine.makeConnection("[Master]", "vu_meter_left", DJCi300.vuMeterUpdateMain);
```
another [Master]->[Main] update needed I think?

BoredGuy1 replied:
Maybe? I was following the wiki, but I guess the wiki might be outdated. It still says to use `VuMeter`, `VuMeterL`, and `VuMeterR`, and Paulus told me those throw deprecated warnings in Zulip.

![image](https://github.com/user-attachments/assets/12e2cc36-6c5e-4104-937c-8eca2e5d0b82)

This is the 2.4 wiki.

Swiftb0y replied:
Many of these changes unfortunately have not made it to the manual page yet. Please try using `[Main]` whenever possible and only revert to `[Master]` when you get errors or broken functionality.

ywwg reviewed (approved):
looking great! approving pending @Swiftb0y 's comments

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -529,27 +523,23 @@ DJCi300.updateToneplayLED = function(value, group, _control) {
-        slicerSizeConnection1 = engine.makeConnection("[Channel1]", "beatloop_size", DJCi300.slicerChangeSize);
-        slicerTrackConnection1 = engine.makeConnection("[Channel1]", "LoadSelectedTrack", DJCi300.slicerLoadTrack);
-        slicerBeatConnection1 = engine.makeConnection("[Channel1]", "beat_distance", DJCi300.slicerCountBeat);
+        DJCi300.slicerConnections[size][index] = engine.makeConnection("[Channel1]", "beatloop_size", DJCi300.slicerChangeSize);
+        DJCi300.slicerConnections[load][index] = engine.makeConnection("[Channel1]", "LoadSelectedTrack", DJCi300.slicerLoadTrack);
+        DJCi300.slicerConnections[beat][index] = engine.makeConnection("[Channel1]", "beat_distance", DJCi300.slicerCountBeat);
```
this will look up strings in a variable called `size` (which is undefined). Instead do 
```suggestion
        DJCi300.slicerConnections.size[index] = engine.makeConnection("[Channel1]", "beatloop_size", DJCi300.slicerChangeSize);
        DJCi300.slicerConnections.load[index] = engine.makeConnection("[Channel1]", "LoadSelectedTrack", DJCi300.slicerLoadTrack);
        DJCi300.slicerConnections.beat[index] = engine.makeConnection("[Channel1]", "beat_distance", DJCi300.slicerCountBeat);
```

BoredGuy1 committed:
Replaced Master with Main

BoredGuy1 committed:
Fixed some slicer variables

BoredGuy1 committed:
Fixed small channel mistake

ywwg commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -185,11 +376,358 @@ DJCi300.scratchWheel = function(channel, control, value, status, _group) {
+    return samplesPerBeat;
+};
+
+// Helper function that calculates current position of play indicator in samples
+DJCi300._currentPosition = function(deck) {
+    const beatClosest = engine.getValue(`[Channel${  deck  }]`, "beat_closest");
+    let beatDistance = engine.getValue(`[Channel${  deck  }]`, "beat_distance");
+
+    // Map beatDistance so that it scales from 0 to .5, then -.5 to 0
+    beatDistance = (beatDistance > .5) ? (beatDistance - 1) : beatDistance;
+    // Adjust beatClosest and return
+    return (DJCi300._samplesPerBeat(deck) * beatDistance) + beatClosest;
+};
```
we already have beat_active: https://github.com/mixxxdj/mixxx/blob/028e3ff4ee0ca464c7c47faa89a1c12a3e99b4e9/src/engine/controls/clockcontrol.cpp#L54

BoredGuy1 replied:
You're right, `track_samples` * `playposition` works perfectly fine. Now that I think about it, there's not much of a reason for the helper function to exist at all - it's only used once anyway, and I can just substitute that line with `engine.getValue(group, "track_samples") * engine.getValue(group, "playposition");`.

Swiftb0y commented on res/controllers/Hercules-DJControl-Inpulse-300-script.js:
```diff
@@ -185,11 +376,358 @@ DJCi300.scratchWheel = function(channel, control, value, status, _group) {
-        "[Channel" + channel + "]", "jog", interval * DJCi300.bendScale);
+// Helper function that calculates samples per beat
+DJCi300._samplesPerBeat = function(deck) {
+    const sampleRate = engine.getValue(`[Channel${  deck  }]`, "track_samplerate");
+    const bpm = engine.getValue(`[Channel${  deck  }]`, "local_bpm");
+    // For some reason, multiplying by 60 makes the size 1/2 as large as it's supposed to be
+    // Hence, we multiply by 120 instead
```
I don't quite understand. Are you arguing that returning 2x the audio sampling frequency is correct behavior of the CO or that its a bug?

BoredGuy1 replied:
I originally put the comment there because I wasn't sure why minutes per beat had to be multiplied by 2*60 instead of just 60 to get the "correct" seconds per beat value. But if it's the sample rate instead of the framerate, then that makes sense. I'll change the comment to reflect that.

BoredGuy1 committed:
replaced decks with groups

BoredGuy1 committed:
reworked toneplay, updated comment about framerate

BoredGuy1 committed:
removed useless helper function

BoredGuy1 committed:
reworked slicer a bit

BoredGuy1 committed:
Packed Slicer into components

BoredGuy1 committed:
Increased beatmatch tolerances by a bit

BoredGuy1 committed:
Increased beatmatch tolerances by a bit

Fixed minor error on comment

Reverted beatmatch code slightly

Tested code, fixed ESLint issues

Fixed even more ESLint issues

3rd time's the charm

BoredGuy1 committed:
Fixed minor error on comment

BoredGuy1 committed:
Reverted beatmatch code slightly

Swiftb0y reviewed (approved):
Thank you. I don't think I can make any more informed comments. I trust you to have the mapping tested thoroughly as I can't confirm whether its working myself. @ywwg merge?

ywwg commented:
@BoredGuy1 can you fix the lint issues identified by the CI?  they look like they are small things

ywwg reviewed (changes requested):
address lint issues

BoredGuy1 committed:
Tested code, fixed ESLint issues

ywwg commented:
sorry just one more set of lint!

```
diff --git a/res/controllers/Hercules-DJControl-Inpulse-300-script.js b/res/controllers/Hercules-DJControl-Inpulse-300-script.js
index 2b48a5f917..507d308c25 100644
--- a/res/controllers/Hercules-DJControl-Inpulse-300-script.js
+++ b/res/controllers/Hercules-DJControl-Inpulse-300-script.js
@@ -585,7 +585,7 @@ DJCi300.Deck = function(deckNumber) {
                         this.slicerUpdateLED(group);
                     }
                 }.bind(this),
-            })
+            });
         }
     }
     // This function will count beats and move the Slicer section forward when needed
@@ -608,7 +608,7 @@ DJCi300.Deck = function(deckNumber) {
 
         // If in slicer mode (not slicer loop mode), check to see if the slicer section needs to be moved
  if (DJCi300.padMode[group] === DJCi300.padModeSlicer) {
  -    
  +
               // If slicerBeat is 8, move the slicer section forward
               if (beat > 7) {
```

BoredGuy1 committed:
Fixed even more ESLint issues

BoredGuy1 commented:
Don't worry - I already know my code style is messy :P

Let me know if there's any more lint issues - I'll gladly patch them.

BoredGuy1 committed:
3rd time's the charm

BoredGuy1 commented:
Let me see if I can figure it out.

BoredGuy1 committed:
Merge branch 'main' of https://github.com/BoredGuy1/mixxx

BoredGuy1 commented:
Never mind, I don't think I did it right. If you can squash down the commits on your end, that would be great.

ywwg commented:
alternative squashed version: https://github.com/mixxxdj/mixxx/pull/14051 @Swiftb0y what do you prefer

Swiftb0y commented:
Lets continue on #14051. Thank you so far @BoredGuy1.
